### ISSUE TITLE
can you please add one diesel + postgres example?

### ISSUE THREAD

**Issue by kushwahashiv** [tokio-rs/axum#475]
specially one covering transactions. how to achieve that using axum.

**Comment by davidpdrsn** [comment]
I haven't used diesel in a very long time and never from async code. So I'm not sure what the best way to do it is. I guess `tokio::task::spawn_blocking` and you can use diesel as par usual but I'm not sure about transactions.

**Comment by Vesafary** [comment]
Maybe this crate will help you: https://docs.rs/bb8-diesel/0.2.1/bb8_diesel/
Here's an example of the crate: https://github.com/overdrivenpotato/bb8-diesel/blob/master/examples/usage.rs

And together with this Axum example you should be able to make it work :)
https://github.com/tokio-rs/axum/blob/main/examples/tokio-postgres/src/main.rs

**Comment by dark-fusion** [comment]
I haven't used `diesel` in a very long time, either, but I understand how to put together an example for it.

If there's demand and you'd like a hand with this, I can create an example that shows axum with diesel + postgresql (including transactions).

**Comment by davidpdrsn** [comment]
That would be great! Depending on the size of the example we could also consider linking to it from [here](https://github.com/tokio-rs/axum/blob/main/ECOSYSTEM.md).

**Comment by dark-fusion** [comment]
Awesome 🙂! I'll get started on it as soon as possible. I don't anticipate that it will take too long to finish. I'll try to keep the example small, but not at the expense of clarity.

I'm okay with whichever location you think suits the example best (examples / community projects / etc).

**Comment by davidpdrsn** [comment]
Sounds good! Just make a PR with our stuff and then we can discuss it 😊

**Comment by dark-fusion** [comment]
@kushwahashiv I understand you would like an example that includes `diesel` and SQL transactions. Are you interested in `diesel` with the `postgres` feature? I just want to clarify before I finish putting things together.

I'm operating on the assumption that "diesel + postgres" == `diesel` w/ `postgres` feature enabled. If there are additional constraints or requirements, please let me know.

Thanks

**Comment by kushwahashiv** [comment]
> @kushwahashiv I understand you would like an example that includes `diesel` and SQL transactions. Are you interested in `diesel` with the `postgres` feature? I just want to clarify before I finish putting things together.
> 
> I'm operating on the assumption that "diesel + postgres" == `diesel` w/ `postgres` feature enabled. If there are additional constraints or requirements, please let me know.
> 
> Thanks

This would be very helpful to understand how the transactions work.

**Comment by fxposter** [comment]
Did anyone check this https://github.com/mehcode/tokio-diesel?

**Comment by rich-murphey** [comment]
> Did anyone check this https://github.com/mehcode/tokio-diesel?

Tokio-diesel runs each query inside a `tokio::task::block_in_place()`:
https://github.com/mehcode/tokio-diesel/blob/f4af42558246ab323600622ba8d08803d3c18842/src/lib.rs#L111

When one knows that a section of code will block the tokio executor, this is a valid workaround.  It moves all other tasks off of the current executor and into a different thread.  The current thread blocks, while executors in other threads are able to run other tasks.  One remaining concern might be that if there are N threads (often one per CPU core)  and there are N such concurrent queries, it could block all of the threads.  So, yes, this is a valid workaround; however, it's still possible to see low CPU utilization when the number of queries reaches the number of threads.  

To summarize, Yes, tokio-diesel can be perform significantly better than diesel alone, especially in terms of reducing latency that would otherwise increase due to blocking.  However, under heavy load, an async db library such as `sqlx` may perform better (and have better CPU utilization).

Just something to consider.  The axum + diesel + postgres example may be helpful regardless, for those that are trying to understand their usage, and under moderate load tokio-diesel should improve latency.

**Comment by kushwahashiv** [comment]
> However, under heavy load, an async db library such as `sqlx` may perform better (and have better CPU utilization).

My earlier code before migration was using `tokio-diesel` but we no longer use it.

**Comment by lz1998** [comment]
> specially one covering transactions. how to achieve that using axum.

Try [sea-orm](https://github.com/SeaQL/sea-orm). It's an async ORM, and I have written a simple example. [Book Management](https://github.com/lz1998/axum-book-management)

**Comment by davidpdrsn** [comment]
I think I'll close this for now as some good solutions have been posted. Someone is still welcome to contribute an example!

**Comment by lukehsiao** [comment]
There is also an axum example on the SeaORM main repository:

https://github.com/SeaQL/sea-orm/tree/master/examples/axum_example
